### DISCUSSION TITLE
How do I send name of player who joined to Discord?

### DISCUSSION THREAD

**Discussion by Cevap05** [lloesche/valheim-server-docker#519] [General]
I have my valheim.env setup and following instructions.
My goal is to have players who join the server, for Discord to notify it in a channel. The webhook is working for this as:

DISCORD_MESSAGE=Starting Valheim Server!
PRE_BOOTSTRAP_HOOK=curl -sfSL -X POST -H "Content-Type: application/json" -d "{\"username\":\"Valheim\",\"content\":\"$DISCORD_MESSAGE\"}" "$DISCORD_WEBHOOK"

In this case, the message goes through saying the server is starting. But when I do this:

VALHEIM_LOG_FILTER_CONTAINS_Spawned="Got character ZDOID from" \
ON_VALHEIM_LOG_FILTER_CONTAINS_Spawned='{ read l; l=${l//*ZDOID from /}; l=${l// :*/}; msg="Player $l spawned into the world"; curl -sfSL -X POST -H "Content-Type: application/json" -d "{\"username\":\"Valheim\",\"content\":\"$msg\"}" "$DISCORD_WEBHOOK"; }'

I get an error saying the "l" and "msg" variables are empty. Can someone help with this thanks!

**Comment by schildbach** [comment]
On top of that, if you only match on "Got character ZDOID from" you will get your message too each time someone dies.

**Comment by PCNicholasAngel** [comment]
I did something a little different using a python script.

Using Docker compose add these to start up:
```
volumes:
  - /location/on/dockerhost:/pythonscripts

environment:
  - POST_BOOTSTRAP_HOOK=apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y python3-pip && pip3 install requests
  - VALHEIM_LOG_FILTER_CONTAINS_Connected=Got handshake from client
  - ON_VALHEIM_LOG_FILTER_CONTAINS_Connected=read l; l=$${l//*from client /}; l=$${l// :*/}; python3 /pythonscripts/playerconnected.py $$l
```
playerconnected.py file:
```
import requests 
import xml.etree.ElementTree as ET 
import sys

steamid_url = 'http://steamcommunity.com/profiles/' + sys.argv[1] + '/?xml=1'
steamidlookup = requests.get(steamid_url)
steamiddata = ET.fromstring(steamidlookup.content)
for steamname in steamiddata.iter('steamID'):  
    playersteamname = steamname.text
    break

discordhook_url = 'https://discord.com/api/webhooks/...'
discord_content = playersteamname + " has connected to the Valheim server."
discord_json = {'username': 'valheim-server',
                'content': discord_content}
discord_post = requests.post(url = discordhook_url, data = discord_json)

print(discord_post)
```


This watches for the 'Got handshake' log with the users Steam ID and then calls Steam and gets the xml output of the users profile, gets their name, then posts the connection message with the Steam name to Discord. You need to update the discordhook_url variable in the python script to your webhook URL.

**Reply by troubleshootme** [reply]
This works well, thank you. I do believe handshake from client occurs before the player authenticates with server, but no big deal. 
I did have to change docker stacks uid and gid to root so it could install dependencies. 

I modified this to send player disconnect events as my player connected hook was already working and since it only reports steam id for disconnects, this python script was exactly what I have been looking for.

Here's what my disconnect hook looks like:
Note:  Make sure the bootstrap_hook below only exists once, just add the 2 below if you already have it
docker-compose.yml
```
  volumes: 
     - /valheim-server/pythonscripts:/pythonscripts
     
  environment:
     - POST_BOOTSTRAP_HOOK=apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y python3-pip && pip3 install requests
     - VALHEIM_LOG_FILTER_CONTAINS_Disconnected=Closing socket
     - ON_VALHEIM_LOG_FILTER_CONTAINS_Disconnected=read l; l=$${l//*Closing socket /}; l=$${l// :*/}; python3 /pythonscripts/playerdisconnect.py $$l
```

playerdisconnect.py:
```
import requests 
import xml.etree.ElementTree as ET 
import sys

steamid_url = 'http://steamcommunity.com/profiles/' + sys.argv[1] + '/?xml=1'
steamidlookup = requests.get(steamid_url)
steamiddata = ET.fromstring(steamidlookup.content)
for steamname in steamiddata.iter('steamID'):  
    playersteamname = steamname.text
    break

discordhook_url = 'your_discord_webhook'
discord_content = playersteamname + " has disconnected from the Valheim server."
discord_json = {'username': 'Valheim',
                'content': discord_content}
discord_post = requests.post(url = discordhook_url, data = discord_json)

print(discord_post)

```
Thanks @PCNicholasAngel !
